### PR TITLE
Let /tavily be served by a model's own web search

### PR DESCRIPTION
**Pull request by giswqs**
Prototype of a second backend for the `/tavily` search route, selected by an environment variable. Tavily stays the default; setting `SEARCH_BACKEND=messages` serves the same route from an Anthropic-messages endpoint that exposes the server-side `web_search` tool (e.g. a self-hosted [cli-proxy-api](https://github.com/router-for-me/CLIProxyAPI)), removing the third-party search dependency.

The route keeps its name **and its Tavily-shaped response**, so `src/lib/opera/news.ts` and the OPERA plugin need no change.

## Configuration

| Setting | Required | Meaning |
| --- | --- | --- |
| `SEARCH_BACKEND` | no | `tavily` (default) or `messages`; anything else is treated as `tavily` |
| `SEARCH_MESSAGES_URL` | for `messages` | base URL; `/v1/messages` is appended |
| `SEARCH_MESSAGES_API_KEY` | for `messages` | bearer token |
| `SEARCH_MESSAGES_MODEL` | no | defaults to `claude-sonnet-5` |

## Design notes

**The model writes the snippets.** Anthropic returns each hit's page text as opaque `encrypted_content` — only `title`, `url` and `page_age` are readable. Since `news.ts` relies on `snippet` to ground quantified figures, the model is asked to return the whole envelope as JSON including a per-source extract.

**Hallucinated links cannot reach the client.** Cited URLs are intersected with the set the search actually returned; an entry with a URL search never produced is dropped.

**Malformed replies degrade, they don't fail.** If the model returns prose instead of JSON, the handler falls back to the raw search hits so the caller still gets citable sources, with the prose as `answer`.

**`page_age` arrives as the string `"None"`** for many pages; that sentinel is dropped rather than surfaced as a date.

**Upstream error bodies are not forwarded** — a non-OK upstream becomes a flat `502`, keeping the client's error contract stable.

## Verification

14 new unit tests, plus the 6 existing Tavily tests still pass; `tsc` and `eslint` clean.

Live end-to-end against a real cli-proxy-api, both topics:

| query | results | sample |
| --- | --- | --- |
| Valencia flood Oct 2024 (`general`) | 4 | Wikipedia — `"Deaths: 237. Property damage: ~3.5 billion euros…"` |
| Texas Hill Country July 2026 (`news`, 90d) | 3 | CNN — `"Parts of Texas Hill Country received over 2 feet of rain…"` |

Real URLs, real dates, quantified figures in the snippets.

## Caveats

- Search + synthesis is slower and far more token-hungry than a Tavily call — tens of thousands of tokens per search against whatever account backs `SEARCH_MESSAGES_URL`. The timeout is therefore 90s, versus Tavily's 30s header bound.
- `published_date` formats vary (`2026-03-03`, `July 17, 2026`, `1 month ago`). `NewsResult.date` is documented as "ISO or free text", so this is within contract, but it is looser than Tavily's.
- New settings are declared on `Env` in source rather than `wrangler.jsonc`: the `secrets` block only supports `required`, which would warn on every deployment that stays on Tavily.

## Summary by CodeRabbit

- **New Features**
  - Added an alternative AI-powered web search backend for `/tavily`.
  - Backend, endpoint, API key, and model settings are configurable; Tavily remains the default.
  - Responses retain the existing result format with validated citations and fallback results.
  - Added request limits, timeout handling, and clearer upstream error responses.

- **Documentation**
  - Documented configuration, defaults, compatibility, and backend differences.

- **Tests**
  - Added comprehensive coverage for selection, parsing, validation, limits, fallbacks, and edge cases.

### CODE REVIEW

File: workers/ai-proxy/src/index.ts [hunk 1]
```diff
@@ -144,6 +164,155 @@ export function buildTavilySearchRequest(
...
+          ? entry.published_date.trim()
+          : hit?.published_date;
+      return {
+        title: typeof entry.title === "string" && entry.title ? entry.title : (hit?.title ?? ""),
+        url,
+        content: typeof entry.content === "string" ? entry.content : "",
+        ...(published ? { published_date: published } : {}),
+      };
+    })
+    .filter((entry): entry is MessagesSearchPayload["results"][number] => entry !== undefined);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply workers/ai-proxy/src/index.ts]: Done in 2bb49297 — you're right, and it contradicted what the README and PR description claimed. The map called `searched.get(url)` but only used the result for fallback title/date, so any non-empty `entry.url` survived. It now returns `undefined` when the URL is absent from the search hits. Added two tests: a fabricated URL alongside a real one (only the real one survives), and the all-invented case (falls back to raw hits). Also re-ran a live search to confirm the intersection doesn't drop legitimate citations — full snippets still come through.

File: workers/ai-proxy/src/index.ts [hunk 2]
```diff
@@ -144,6 +164,155 @@ export function buildTavilySearchRequest(
...
+  // has sources to cite, with the model's prose as the answer.
+  return {
+    results: [...searched.entries()].map(([url, hit]) => ({
+      title: hit.title,
+      url,
+      content: "",
+      ...(hit.published_date ? { published_date: hit.published_date } : {}),
+    })),
+    answer: answer ?? (textParts.join("").trim() || undefined),
+  };
```
[the remark this answers is not in the document]
Comment (giswqs) [reply workers/ai-proxy/src/index.ts:330]: Done in 2bb49297 — extracted the clamp as `searchResultLimit()` so the prompt and the response cap share one bound, and sliced both the claimed list and the fallback branch to it. The handler passes the client's limit into `parseMessagesSearchResponse`. Tests cover both branches plus the clamp itself.

File: workers/ai-proxy/src/index.ts [hunk 3]
```diff
@@ -319,6 +498,104 @@ async function proxyTavily(request: Request, env: Env, origin: string | null): P
...
+      signal: controller.signal,
+    });
+  } catch (error) {
+    if (controller.signal.aborted) {
+      return jsonError("Search request timed out", 504, responseHeaders(origin));
+    }
+    throw error;
+  } finally {
+    clearTimeout(deadline);
+  }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply workers/ai-proxy/src/index.ts]: Done in 2bb49297 — correct, and it made the comments at lines 32-34 and 544-545 false rather than merely optimistic. The `finally` fired when `fetch` resolved at the headers, leaving `upstream.json()` unbounded with the signal already disarmed. Moved `clearTimeout` to an outer `finally` that encloses the body read, and an abort during the read now reports 504 instead of falling into the unreadable-response 502.

File: workers/ai-proxy/src/index.ts [hunk 4]
```diff
@@ -144,6 +164,167 @@ export function buildTavilySearchRequest(
...
+      if (!hit) return undefined;
+      const published =
+        typeof entry.published_date === "string" && entry.published_date.trim()
+          ? entry.published_date.trim()
+          : hit?.published_date;
+      return {
+        title: typeof entry.title === "string" && entry.title ? entry.title : hit.title,
+        url,
+        content: typeof entry.content === "string" ? entry.content : "",
+        ...(published ? { published_date: published } : {}),
```
[the remark this answers is not in the document]
Comment (giswqs) [reply workers/ai-proxy/src/index.ts:305]: Done in ac850cd3 — agreed. An entry with a blank snippet grounds nothing and doesn't meet the schema the prompt asks for, yet counted as a non-empty result and so suppressed the raw-hit fallback. Entries whose `content` is missing or whitespace-only are now dropped, so a mixed reply keeps just the usable citations and an all-blank reply falls back to the search hits. Added both cases as tests, and re-ran a live search to confirm the guard doesn't drop legitimate results (4 and 3 results respectively, unchanged).
